### ISSUE TITLE
Heartbeat regressions in v4.10.0

### ISSUE THREAD

**Issue by adamantike** [robtaussig/react-use-websocket#250]
We have upgraded to `4.10.0`, and started getting a different behavior than what we had with `4.9.0`. I'm listing here what I could investigate, and am available for further testing if needed!

For reference, a simplification of our implementation looks like this:

```typescript
useWebSocket<T>(
  url,
  {
    queryParams: requiresAuth ? { credentials: `Bearer ${token}` } : undefined,
    shouldReconnect: () => true,
    reconnectAttempts: 10,
    filter: event => event.data !== '"pong"',
    heartbeat: {
      message: JSON.stringify({ action: EWebsocketActions.HEARTBEAT }),
      returnMessage: 'pong',
      interval: 30_000,
    },
    ...props.options,
  },
  shouldConnect,
);
```

1. Only the first connection is able to send a heartbeat message, and correctly receives the `pong` response. This is triggered 30 seconds after the connection was initialized. During that time, the only message the connection received was that `pong`. At the 60 second mark, the library logs the following warning, and closes the connection: `Heartbeat timed out, closing connection, last message was seen 60200ms ago`. This seems to indicate that the `pong` received message is not actually updating the `lastMessageTime` variable.
2. Reconnection is successful, but seems to **not** reset the `lastMessageTime`, so future connections are **not** triggering the heartbeat anymore, and we end up with a snowball of closed connections (none of them received a message from the backend -which is expected-, but also didn't trigger a single heartbeat message). Because `lastMessageTime` is not reset, logs keep showing an increasing amount of time without received messages: `Heartbeat timed out, closing connection, last message was seen 2111790ms ago`
3. While checking out the implementation from #246, there's a mismatch between the `shared` mode and the regular mode. For the regular mode, `lastMessageTime` is updated **before** the `filter` is evaluated, while it's the other way around for the shared mode.

**Comment by w666** [comment]
Hi @adamantike, I will have a look on this.

**Comment by w666** [comment]
Should be fixed, @adamantike and @robtaussig could you please review PR?

**Comment by robtaussig** [comment]
Approved and released as [4.10.1](https://www.npmjs.com/package/react-use-websocket/v/4.10.1).

Thanks for the report @adamantike and for the quick fix @w666 !
